libs/community/langchain_google_community/bq_storage_vectorstores/bigquery.py

```diff
@@ -0,0 +1,489 @@
+                    self.text_embedding_field,
+                    self.content_field,
+                ]:
+                    metadata[field] = row[field]
+            metadata["__id"] = row[self.doc_id_field]
+            doc = Document(page_content=row[self.content_field], metadata=metadata)
+            docs.append(doc)
+        return docs
+
+    def _initialize_bq_vector_index(self) -> Any:
+        """
+        A vector index in BigQuery table enables efficient
+        approximate vector search.
+        """
+        if self._have_index or self._creating_index:
+            return
+
+        table = self._bq_client.get_table(self._full_table_id)  # type: ignore[union-attr]
+        if (table.num_rows or 0) < MIN_INDEX_ROWS:
+            self._logger.debug("Not enough rows to create a vector index.")
+            return
+
+        if datetime.utcnow() - self._last_index_check < INDEX_CHECK_INTERVAL:
+            return
+
+        with _vector_table_lock:
+            self._last_index_check = datetime.utcnow()
+            # Check if index exists, create if necessary
+            check_query = (
+                f"SELECT 1 FROM `{self.project_id}."
+                f"{self.dataset_name}"
+                ".INFORMATION_SCHEMA.VECTOR_INDEXES` WHERE"
+                f" table_name = '{self.table_name}'"
+            )
+            job = self._bq_client.query(  # type: ignore[union-attr]
+                check_query, api_method=self._bigquery.enums.QueryApiMethod.QUERY
+            )
+            if job.result().total_rows == 0:
+                # Need to create an index. Make it in a separate thread.
+                self._create_bq_index_in_background()
+            else:
+                self._logger.debug("Vector index already exists.")
+                self._have_index = True
+
+    def _create_bq_index_in_background(self) -> None:
+        if self._have_index or self._creating_index:
+            return
+
+        self._creating_index = True
+        self._logger.debug("Trying to create a vector index.")
+        Thread(target=self._create_bq_index, daemon=True).start()
+
+    def _create_bq_index(self) -> None:
+        table = self._bq_client.get_table(self._full_table_id)  # type: ignore[union-attr]
+        if (table.num_rows or 0) < MIN_INDEX_ROWS:
+            return
+
+        index_name = f"{self.table_name}_langchain_index"
+        try:
+            sql = f"""
+                CREATE VECTOR INDEX IF NOT EXISTS
+                `{index_name}`
+                ON `{self._full_table_id}`
+                ({self.text_embedding_field})
+                OPTIONS(distance_type="{self.distance_type}", index_type="IVF")
+            """
+            self._bq_client.query(sql).result()  # type: ignore[union-attr]
+            self._have_index = True
+        except ClientError as ex:
+            self._logger.debug("Vector index creation failed (%s).", ex.args[0])
+        finally:
+            self._creating_index = False
```

Comment:
I notice that the FeatureStore part has  the functionality to wait for things to be ready. 
BQ index is created partially asynchronously. The return of the DDL does not indicate that the index is finished. Users actually need to check the INFORMATION SCHEMA or run a query to know that.

I wonder if we should have the option for users to wait for BQ index to be ready; or document the behavior in the comment or the langchain document.

Reply:
@jinmmin This is a good point. For BQ we retained the functionality of the former BigQueryVectorSearch class.
Having said that, I think the difference with FS is that in BQ it is still possible to query on the full set of data even if the index is not ready. On the contrary, for FS you need the sync process to finish in order to be able to query on the whole dataset.